src/main/java/duke/TasksList.java

```diff
@@ -0,0 +1,166 @@
+            String raw_task_entry = file_scanner.nextLine();
+            String[] task_data_points = raw_task_entry.split("\\|");
+            switch (task_data_points[0]) {
+            case "T":
+                task = new ToDo(task_data_points[2]);
+                if (task_data_points[1].equals("1")) {
+                    task.markAsDone();
+                }
+                tasks.add(task);
+                break;
+            case "D":
+                task = new Deadline(task_data_points[2], task_data_points[3]);
+                if (task_data_points[1].equals("1")) {
+                    task.markAsDone();
+                }
+                tasks.add(task);
+                break;
+            case "E":
+                task = new Event(task_data_points[2], task_data_points[3]);
+                if (task_data_points[1].equals("1")) {
+                    task.markAsDone();
+                }
+                tasks.add(task);
+                break;
+            default:
+                throw new DukeException();
+            }
+        }
+    }
+
+    // Saves all tasks to ./data/duke.txt
+    public void saveTasks() throws IOException{
+        FileWriter file_writer = new FileWriter(FILE_DIR + "/" + FILE_NAME);
+        StringBuilder output_string = new StringBuilder();
+        for (Task task : tasks) {
+            output_string.append(task.formatForDataStore());
+        }
+        file_writer.write(output_string.toString());
+        file_writer.close();
+    }
+
+    public int getSize() {
+        return tasks.size();
+    }
+
+    public String getTaskString(int taskIndex) {
+        Task task = tasks.get(taskIndex);
+        return task.toString();
+    }
+
+    public void addToDo(String rawInput) {
+        try {
+            String description = rawInput.substring(LENGTH_OF_TODO);
+            tasks.add(new ToDo(description));
+            System.out.println("\tGot it. I've added this task:");
+            System.out.println("\t\t" + this.getTaskString(this.getSize() - 1));
+            System.out.println("\tNow you have " + this.getSize() + " tasks in the list.");
+        } catch (StringIndexOutOfBoundsException e) {
+            System.out.println("\tOOPS!!! The description of a todo cannot be empty.");
+        }
+    }
+
+    public void addEvent(String rawInput) {
+        try{
+            int idx = rawInput.indexOf("/at");
+            String description = rawInput.substring(LENGTH_OF_EVENT, idx-1);
+            String at = rawInput.substring(idx+LENGTH_OF_AT);
+            tasks.add(new Event(description, at));
+            System.out.println("\tGot it. I've added this task:");
+            System.out.println("\t\t" + this.getTaskString(this.getSize() - 1));
+            System.out.println("\tNow you have " + this.getSize() + " tasks in the list.");
+        } catch (StringIndexOutOfBoundsException e) {
+            System.out.println("\tWrong deadline format! Please follow `deadline <description> /at <at>`.");
```

Comment:
Avoid the use of magic strings. In this case, the error message can be extracted into a constant variable with a suitable name.